apollo-router/src/plugins/telemetry/config_new/cost/mod.rs

```diff
@@ -0,0 +1,358 @@
+#[derive(Default)]
+pub(crate) struct CostInstruments {
+    /// A histogram of the estimated cost of the operation using the currently configured cost model
+    cost_estimated: Option<
+        CustomHistogram<
+            supergraph::Request,
+            supergraph::Response,
+            SupergraphAttributes,
+            SupergraphSelector,
+        >,
+    >,
+
+    /// A histogram of the actual cost of the operation using the currently configured cost model
+    cost_actual: Option<
+        CustomHistogram<
+            supergraph::Request,
+            supergraph::Response,
+            SupergraphAttributes,
+            SupergraphSelector,
+        >,
+    >,
+    /// A histogram of the delta between the estimated and actual cost of the operation using the currently configured cost model
+    cost_delta: Option<
+        CustomHistogram<
+            supergraph::Request,
+            supergraph::Response,
+            SupergraphAttributes,
+            SupergraphSelector,
+        >,
+    >,
+}
+
+impl Instrumented for CostInstruments {
+    type Request = supergraph::Request;
+    type Response = supergraph::Response;
+
+    fn on_request(&self, request: &Self::Request) {
+        if let Some(cost_estimated) = &self.cost_estimated {
+            cost_estimated.on_request(request);
+        }
+        if let Some(cost_actual) = &self.cost_actual {
+            cost_actual.on_request(request);
+        }
+        if let Some(cost_delta) = &self.cost_delta {
+            cost_delta.on_request(request);
+        }
+    }
+
+    fn on_response(&self, response: &Self::Response) {
+        if let Some(cost_estimated) = &self.cost_estimated {
+            cost_estimated.on_response(response);
+        }
+        if let Some(cost_actual) = &self.cost_actual {
+            cost_actual.on_response(response);
+        }
+        if let Some(cost_delta) = &self.cost_delta {
+            cost_delta.on_response(response);
+        }
+    }
+
+    fn on_error(&self, error: &BoxError, ctx: &Context) {
+        if let Some(cost_estimated) = &self.cost_estimated {
+            cost_estimated.on_error(error, ctx);
+        }
+        if let Some(cost_actual) = &self.cost_actual {
+            cost_actual.on_error(error, ctx);
+        }
+        if let Some(cost_delta) = &self.cost_delta {
+            cost_delta.on_error(error, ctx);
+        }
+    }
+}
+
+#[derive(Deserialize, JsonSchema, Clone, Debug, PartialEq)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub(crate) enum CostValue {
+    /// The estimated cost of the operation using the currently configured cost model
+    Estimated,
+    /// The actual cost of the operation using the currently configured cost model
+    Actual,
+    /// The delta between the estimated and actual cost of the operation using the currently configured cost model
+    Delta,
+    /// The result of the cost calculation. This is the error code returned by the cost calculation.
+    Result,
```

Comment:
Nit: the naming of this is a bit confusing to me because it's very similar `CostResult` and `CostResult.result`. Is there a case where the calculation succeeds and this `Result` is returned? Or is this purely for error values?

Reply:
[a28b9ce](https://github.com/apollographql/router/pull/4990/commits/a28b9cefa3dfb1dc3c7ea281a75abbfae1ded8cd)

Reply:
Renamed